Use rsync to copy pack contents

cognifloyd opened the pull request:
Closes #245

cognifloyd committed:
Use rsync to copy pack contents

cognifloyd commented on templates/_helpers.tpl:
```diff
@@ -343,9 +343,14 @@ Merge packs and virtualenvs from st2 with those from st2packs images
+        rsync -a /opt/stackstorm/packs/. /opt/stackstorm/packs-shared &&
+        rsync -a /opt/stackstorm/virtualenvs/. /opt/stackstorm/virtualenvs-shared;
+      else
+        /bin/cp -aR /opt/stackstorm/packs/. /opt/stackstorm/packs-shared &&
```
`-a` is equivalent to `-dR --preserve=all`:
- `-d`: no dereference (preserve links)
- `-R`: copy directories recursively
- `--preserve=all`: preserve file attributes
  - `mode`
  - `ownership` for user and group
  - `timestamps`
  - `links` for hard links
  - `context` for security context
  - `xattr` for extended attributes

Of these we need `-dR --preserve=mode` (we need mode to maintain the execute bits of action files). Depending on which user is copying the file, we might also need `--preserve=ownership`.

cognifloyd replied:
I think I'll push a commit that switches to `-RP --preserve=mode,timestamps,links,xattr` based on https://github.com/StackStorm/stackstorm-k8s/compare/rsync...Stealthii:stackstorm-k8s:rsync

cognifloyd reviewed (commented):
Either we do something like this, or we make the command configurable via values.

Maybe `.utility.rsync = "rsync -rlp"`. That would mean the user would have to update this if they needed to switch to `"cp"`. Or we keep both, but allow specifying the args.
```yaml
utility:
  rsync: "rsync -a"
  cp: "cp -aR"
```

```yaml
utility:
  rsync: "rsync -rlE"
  cp: "cp -dR --preserve=mode"
```

The gotcha with making it configurable is that people could forget to include recursive, symlinks, and preserving execute bits.

Stealthii commented:
In an attempt to solve the ownership issue that blocks #400, we should only preserve the file attributes that matter for this purpose.

I've updated a fork of this branch with proposed changes that once verified, could be brought into this PR.

https://github.com/StackStorm/stackstorm-k8s/compare/rsync...Stealthii:stackstorm-k8s:rsync

cognifloyd commented on templates/_helpers.tpl:
```diff
@@ -343,9 +343,14 @@ Merge packs and virtualenvs from st2 with those from st2packs images
-      /bin/cp -aR /opt/stackstorm/virtualenvs/. /opt/stackstorm/virtualenvs-shared
+    - >
+      if command rsync; then
+        rsync -a /opt/stackstorm/packs/. /opt/stackstorm/packs-shared &&
```
I think I'll push a commit that switches to `-rlptD` based on https://github.com/StackStorm/stackstorm-k8s/compare/rsync...Stealthii:stackstorm-k8s:rsync

cognifloyd committed:
Don't preserve ownership when copying pack contents (#418)

This change aims to preserve file attributes when copying via rsync or
cp, without ownership to enable environments where containers are ran
without root privileges.

Co-authored-by: Daniel Porter <dporter@proofpoint.com>

cognifloyd committed:
More rsync/cp cleanups

cognifloyd committed:
Merge branch 'master' into rsync
